plugin/src/main/java/com/google/errorprone/xplat/checker/J2objcMethodName.java

```diff
@@ -0,0 +1,326 @@
+import com.google.errorprone.matchers.Description;
+import com.google.errorprone.matchers.Matcher;
+import com.google.errorprone.matchers.Matchers;
+import com.google.errorprone.matchers.MethodVisibility.Visibility;
+import com.google.errorprone.util.ASTHelpers;
+import com.sun.source.tree.AnnotationTree;
+import com.sun.source.tree.ClassTree;
+import com.sun.source.tree.CompilationUnitTree;
+import com.sun.source.tree.MethodTree;
+import com.sun.source.tree.Tree;
+import com.sun.source.tree.VariableTree;
+import com.sun.tools.javac.code.Symbol;
+import com.sun.tools.javac.code.Symbol.ClassSymbol;
+import com.sun.tools.javac.code.Type;
+import java.util.ArrayList;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.Optional;
+import javax.lang.model.type.TypeKind;
+
+/**
+ * Checks for Java methods that would become unreasonably long when translated by J2ObjC into a
+ * function. If a method has 6+ parameters and its name would be 80+ characters, a warning is given.
+ * By default, any method which would have a name with 300+ characters is warned against. This can
+ * be changed with the command line argument {@code -XepOpt:J2ObjCMethodName:MethodNameLength=X}
+ * where X is the minimum number of characters to warn on. For example, {@code
+ * -XepOpt:J2ObjCMethodName:MethodNameLength=100} will warn against all methods that would result in
+ * 100 or more characters after being translated to an Objective-C function.
+ */
+@BugPattern(
+    name = "J2ObjCMethodName",
+    summary = "Warns against long J2ObjC translated methods.",
+    explanation =
+        "J2ObjC translates Java methods in a specific format that can lead to unreasonably long"
+            + " Objective-C functions when the Java method has many parameters and/or uses long"
+            + " types. This checker calls out problematic methods and offers a fix.",
+    severity = WARNING)
+public class J2objcMethodName extends BugChecker implements MethodTreeMatcher,
+    ClassTreeMatcher, CompilationUnitTreeMatcher {
+
+  private final Map<String, String> foundObjcClassNames = new HashMap<>();
+
+  private final Map<String, Integer> usedMethodNames = new HashMap<>();
+
+  private String packageAnnotation;
+
+  private int methodNameLength = 300;
+
+  private static final Matcher<MethodTree> MATCHER =
+      Matchers.anyOf(
+          Matchers.isStatic(),
+          Matchers.methodHasVisibility(Visibility.PRIVATE)
+      );
+
+  private static final Matcher<Tree> OBJC_NAME_MATCHER =
+      Matchers.hasAnnotation("com.google.j2objc.annotations.ObjectiveCName");
+
+  private Description genDescription(String message, MethodTree tree) {
+    return buildDescription(tree)
+        .setMessage(message)
+        .addFix(SuggestedFix.builder()
+            .addImport("com.google.j2objc.annotations.ObjectiveCName")
+            .prefixWith(tree, String.format("@ObjectiveCName(\"%s\")\n",
+                getMethodName(tree.getName().toString())))
```

Comment:
Is it worth checking for conflicts with existing ObjectCName tags (or other method names)

Reply:
I do make sure that names I generate do not conflict with themselves, but I don't think there is a good way to make sure my generated names do not conflict anywhere, as that would require checking all of the method names once before letting my check run. The name I generate is simply the name of the java method, and if there are multiple methods with the same name, subsequent methods gain a number after the java method name. The internal document on the subject it also mentions that simplifying method names "should be safe nonetheless due to link-time duplicate symbol checking". If this is important to have, I can try and figure out a way to do it.

Reply:
I think it's safe not to do this check. The C linker is already a guard rail like David pointed out.